src/dict.h

```diff
@@ -53,6 +53,7 @@ typedef struct dictEntry {
         double d;
     } v;
     struct dictEntry *next;
+    void *metadata[];
```

Comment:
Yes, and I'd prefer to keep the char array, and the NEW version is what you had in the original dictMetadata macro. The only reason I changed from `char metadata[]` to `void *metadata[]` is that I got a warning/error "dereferencing type-punned pointer will break strict-aliasing rules" on build-ubuntu-old (and another build, which I think is using some old GCC version) when casting the `char metadata[]` as `(dictEntry**)(void*)metadata`. I get the same warning with and without the middle `void*` cast) gives this warning, so changing the macro from OLD to NEW doesn't make any difference here.

The GCC version seems to dislike casting char array to void pointer, but it seems any pointer can be cast to char array without problems (as we do when we call memset on the metadata after allocation). If you have another solution to this, I'm happy to change back to char array.

Note: Even if it's now a `void* metadata[]`, it's still possible to use it as an arbitrary number of bytes and cast it to `char*` when using it (even though it seems wrong).

Reply:
I have `char metadata[]` in another branch, so you can see what fails in this build: https://github.com/zuiderkwast/redis/runs/3309598271.

Reply:
I think the compiler is warning that `char[]` is not inherently aligned (it's aligned to a multiple of 1).  When we cast to a pointer (even when trying to misdirect the compiler with `void*`) it knows that the `char[]` is not 8-byte aligned.

In our case, we know that it's aligned because the data is placed immediately after another pointer.  So the code should work safely, though the compiler warning is certainly annoying.

I've seen some suggestions to create a union.  The union would server to force alignment with the unioned type.  But I agree that this looks messy.

The cleanest option might be the one you proposed `void *metadata[]`.  By making it an array of pointers, the compiler knows that the array is aligned.  If this is the case, I think it deserves a comment.  Something like:
```
// an arbitrary number of characters, starting at a pointer-aligned address
```

One thing that I'm not sure about is if this warning would get generated for other types given an array of pointers.  For instance, if we define:
```
typedef struct clusterDictEntryMetadata {
    int32_t intval;    // a 4 byte int
    // (the compiler will insert 4 bytes of padding here)
    dictEntry *prev;            /* Prev entry with key in the same slot */
    dictEntry *next;            /* Next entry with key in the same slot */
} clusterDictEntryMetadata;
```
Does this now generate an error when we try to access `intval`?

Reply:
Thanks for digging into this @JimB123. It make sense that it's an alignment related problem and that newer GCC versions correctly finds that it's already pointer aligned. Another possibly is that type-punning is simply allowed on a `void*` by C99 so the compiler just has to accept it.

I tried to add an int32_t member as you said and there's no build warning. Here's the branch: https://github.com/zuiderkwast/redis/tree/try-non-pointer-sized-member-in-pointer-aligned-dict-entry-metadata

Reply:
I'm ok with the change then.  A comment should clarify what's going on.  Your proposal seems like the simplest solution.

Reply:
Good. I tried `union { char bytes[]; void *pointers[] } metadata;` too but it's not allowed to have arrays without size in a union.